### ISSUE TITLE
Ensure cross-node calls have a timeout

### ISSUE THREAD

**Issue by michaelklishin** [rabbitmq/rabbitmq-server#166]
This is related to #62.

We currently do remote node calls with the timeout of `infinity`. This is not exactly great from the operations perspective and we should try to have a more sensible timeout in as many places as possible.

**Comment by michaelklishin** [comment]
https://github.com/rabbitmq/rabbitmq-server/issues/248 is now being a part of this.

**Comment by Ayanda-D** [comment]
With all the concerns on implications on service when delegates timing out, or similar elements carrying out cross-node calls, some ideas on how we could actually make this feasible, in much a safer manner. I think our main limitation is the use of the `gen_server` behaviour to implement delegates. We have little control on setting any rules on when, and when not, a timeout should occur. I think re-implementing delegates to retain the same functionality, but using `proc_lib`, could give us more flexibility on adding rules relating to service, prior to a delegate exiting due to a timeout. For example;

- Timeout only if certain a alarm (or control message) from a queue has been received - queues could be updated to send some control message to the invoking delegate when subject to certain conditions
- Timeout if the current delegate's message queue is empty (avoiding problems regarding preceding calls and casts)
- Timeout if the remote node is unreachable and we have no preceding calls/casts

(could be more)

Otherwise, if a timeout exit is going to be detrimental to service, rather the delegate resets it timeout and continues running. Another option is to update the `gen_server2` behaviour, to allow handling of timeout rules from the callback, to determine whether the delegate (or any other `gen_server2` process) terminates, or continues running. 

Just another design option before taking this feature into full swing. 

@michaelklishin @binarin

**Comment by michaelklishin** [comment]
Thanks @Ayanda-D. I will do more research and get back to you.

**Comment by binarin** [comment]
I don't think `gen_server` is a problem - using `noreply` and spawning additional helper processes will allow any timeout policy to be implemented. 

But after looking at `delegate` I have following questions:

1. First stated purpose of `delegate` is optimisation of numerous inter-node calls. But actually this can be solved by using just `rpc:call` on `delegate:safe_invoke` - as `rpc` already performs calls in separate process. Or if `rpc` is going to be performance bottleneck, then something like `gen_rpc` could be used instead.

2. Second stated purpose is preserving ordering of messages. But as it happens based solely based on pid of process invoking `delegate` is it actually needed? At least for calls it happens naturally, because invoking process is naturally blocked on `delegate:call` at this point. So only casts need some additional mechanism for enforcing ordering.

3. Looks like the third purpose is reducing number of inter-node monitors. But does it actualy improve anything? Because `delegate:monitor/2` still involves gen_server call, and death of monitored process still result in sending of inter-node message.

Given all of the above are the separate `delegate` processes actually needed? Isn't it better to remove as many custom code as possible and switch to some readily available thing like `rpc` or `gen_rpc`?

**Comment by binarin** [comment]
One more thing that I can't wrap my head around is the comment at https://github.com/rabbitmq/rabbitmq-server/blob/master/src/delegate.erl#L113 and corresponding code. What's wrong with using timeout here? What wrong order of calls/cast infinite timeout prevents at that point?

**Comment by michaelklishin** [comment]
@binarin spawning a new process that could be used for calls means you have a natural race condition that will screw up ordering.

Delegates primarily exist to avoid doing N identical cross-node calls when a single message is routed to N queues on a remote node. The whole thing may cease to exist when we move to Raft. We can perhaps reimplementing parts of it but any substantial rework would seem like (mostly) a waste of time.

**Comment by binarin** [comment]
I think I'm still missing something regarding ordering and timeout.

Helper process starts at the beggining of `gen_server:multi_call` and it life ends before this `gen_server:multi_call` returns. Helper sends calls in exactly the same order as it happens with `infinity` timeout. So for the outside world there should be no difference at all: request were sent in the same order and there is no extra process hanging around after `gen_server:multi_call` returns.

So where is the race condition?

**Comment by Ayanda-D** [comment]
@binarin regarding the helper process you're referring to, are you implying that we spawn and wait for it complete? Then how would that help in determining whether the delegate continues alive or terminates upon a timeout? Bear in mind we only have 16 reusable delegates to work with, which is why a timeout of `infinity` has been appropriate. Also, having 16 delegates seems to have been a good idea in limiting the number of cross-node calling processes, compared to using `rpc`, which would spawn an intermediary process for each and every request.

If we add a timeout, more requests from channels could be inbound during the delegate's termination. So my problem with `gen_server` is, once we exit and the `terminate` function is called, we have no means of keeping the delegate alive, if for instance, the queue is busy (or for any other rules we define). We can sure spawn a helper process, but what happens to the messages held in the terminating delegate's message queue? And again, all this is for synchronous calls. Does the caller get a timeout exception, whilst the helper process continues with proceedings (if it had been spawned asynchronously)?

**Comment by michaelklishin** [comment]
`delegate` certainly has similarities with `rpc`. Both eventually use the same thing under the hood for multi calls, `gen_server`.

`delegate` uses a pool of processes (16 by default) on every node and hashes the caller (typically a queue or channel process) to find out what process on the other end should perform function application. `rpc` uses a single process locally registered as `rex`. This is hardcoded in the `rpc` module.

For calls specifically, as @binarin points out, on the calling end (RPC client, "first half" in RabbitMQ Erlang client speak), there is no real difference as the middleman process in `rpc` that collects late responses is gone by the time `rpc:[multi]call` returns. However, on the receiving end (RPC server), we are limited by a single named process in `rpc` and a configurable pool in case of `delegate`.

Most other differences are probably a lot less important.For a system such as RabbitMQ, where there are normally many queues and processes, to me it makes sense to use a pool.

**Comment by michaelklishin** [comment]
@Ayanda-D sorry, I re-read your comment about re-implementing delegates using `proc_lib` and not sure I quite get what specific cases make you worried. Can you cover a few scenarios in more detail, please?

**Comment by Ayanda-D** [comment]
@michaelklishin the delegate to use for a `multi_call` is acquired by hashing the caller's pid and count of the number of delegate processes for a particular remote node. So if a delegate is referenced by more than one caller, and thus has some requests pending in its message queue, and times out and terminates, what then happens to remaining messages in its message queue?

**Comment by binarin** [comment]
> For calls specifically, as @binarin points out, on the calling end (RPC client, "first half" in RabbitMQ Erlang client speak), there is no real difference as the middleman process in rpc that collects late responses is gone by the time rpc:[multi]call returns. However, on the receiving end (RPC server), we are limited by a single named process in rpc and a configurable pool in case of delegate.

> Most other differences are probably a lot less important.For a system such as RabbitMQ, where there are normally many queues and processes, to me it makes sense to use a pool.

Yes, but currently delegates in that pool are blocked when they perform `invoke` - they are completely synchronous. So probably some measurements should be made to compare those variants - there is even a chance that using `rpc` will improve performance.

Actually I see following easy improvement for `delegate:call/2`:
1. It should invoke delegate on all nodes simutaneously. Currently calls on local node are done after remote nodes. The easiest way to do this is to use single `gen_server2:multi_call` - both for remote and local nodes.
2. `delegate:call/2` should be supported by separate `handle_call` clause in `delegate` server - to make it as asynchronous as possible - by spawning separate process to collect replies and handle timeouts. Something like this (heavily borrowed from `gen_server:multi_call`) - b
```erlang
handle_call({call, Msg, Grouped, Timeout}, From, State = #state{node = Node}) ->
    Pids = orddict:fetch(Node, Grouped),
    Tag = make_ref(),
    Collector = spawn_link(delegate, call_collector, [Pids, From, Timeout, Tag]),
    %% Here we sending all call requests from delegate process, just to be 100% sure
    %% that message ordering will be preserved as in original delegate implementation.
    %% But reply collection will happen in separate process, which will also handle timeouts.
    [ catch Pid ! {'$gen_call', {Collector, Tag}, Msg} || Pid <- Pids ],
    {noreply, State, hibernate);

call_collector(Pids, From, Timeout) ->
    Monitors = [ erlang:monitor(process, Pid) || Pid <- Pids ],
    TimerId = erlang:start_timer(Timeout, self(), ok),
    Result = rec_pids(Tag, Monitors, TimerId, []),
    gen_server2:reply(From, Result).

%% Collecting successful replies and/or waiting for timeout.
rec_pids(Tag, [], TimerId, Replies) ->
    %% Every pid responded, everything is ok.
    case catch erlang:cancel_timer(TimerId) of
        false ->
            receive
                {timeout, TimerId, _} -> ok
            after 0 -> ok
            end;
        _ -> ok
    end,
    Replies;
rec_pids(Tag, [Monitor | Monitors], TimerId, Replies) ->
    receive
        {{Tag, N}, Reply} -> %% Tag is bound!!!
            %% One more successful reply
            rec_pids(Tag, Monitors, TimerId, [Reply|Replies]);
        {timeout, TimerId, _} ->
            %% We are out of time
            rec_pids_rest(Tag, [Monitor|Monitors], Replies)
    end.


rec_pids_rest(Tag, [Monitor|Monitors], Replies) ->
    receive
        {'DOWN', Monitor, _, _, _} ->
              rec_pids_rest(Tag, Monitors, [bad_rpc | Replies]);
        {{Tag, N}, Reply} -> %% Tag is bound!!!
             rec_pids_rest(Tag, Monitors, [Reply|Replies])
     after 0 ->
         rec_pids_rest(Tag, Monitors, [no_reply | Replies])
     end;
rec_pids_rest(_Tag, [], Replies) ->
     Replies.
```

**Comment by michaelklishin** [comment]
@binarin I find it hard to believe that 1 hardcoded RPC handler process will offer better throughput than 16 (or generally, N) processes. Delegates are used by *more than one channel* in parallel in 99% of realistic workloads. I have a feeling this has been ignored entirely in this conversation.

In any case, this issue is not about improving performance, and the title makes that very clear. Not sure what makes you believe it is.

**Comment by michaelklishin** [comment]
@Ayanda-D OK, I see your point. I think timeouts should throw exceptions in the caller, and the delegate handler processes should not terminate. To draw a parallel with `rpc`, imagine what would happen if the `rex` processes terminated due a timeout ;)

**Comment by binarin** [comment]
> @binarin I find it hard to believe that 1 hardcoded RPC handler process will offer better throughput than 16 (or generally, N) processes.

It depends on type of a workload you have. Because RPC server spawns process for every request and its throughput is in theory is limited only by number of `erlang:spawn_monitor/1` it could do a second. On the other hand `delegate` server is blocked by it's `call` workload. So performance probably degrades at the point where that `call` workload becomes more computationally intensive that 16 spawns.

> Delegates are used by more than one channel in parallel in 99% of realistic workloads.

> I have a feeling this has been ignored entirely in this conversation.

Actually not :smile: I started to look into some ways to implement timeouts, and it looks like that by correctly implementing timeouts increased throughput could also be easily gained. I just forgot to mention that changes I've mentioned in previous message are about both delegate timeouts and improved throughput.

**Comment by michaelklishin** [comment]
I have no objections to making `delegate` (more) asynchronous but it should be a separate issue.

**Comment by binarin** [comment]
So imagine we have following set of processes:
- `Pid_Local_10`, `Pid_Local_20`, `Pid_Local_Infinity` - some gen_server's on local node which take 10, 20 and infinity(faulty process) seconds to process some `gen_server:call`.
- `Pid_Remote_10`, `Pid_Remote_20`, `Pid_Remote_Infinity` - remote gen_servers which take corresponding amount of time to do their job.

### First scenario
```erlang
delegate:call([Pid_Remote_10, Pid_Local_Infinity], do_work)
```
Is such problem also out of scope of this issue? Because `delegate:call` will hang, but it will not involve any cross-node calls.

### Second scenario
Suppose now there is new `delegate:call/3` which accepts timeout as a last argument.
```erlang
delegate:call([Pid_Remote_10, Pid_Remote_20, Pid_Local_10], do_work, 25000)
```
Here we have call with timeout of 25 seconds, and every referenced gen_server can do it's job during that time. But:
- In current implementation local calls are performed only after remote ones are done. And call to `Pid_Local_10` will timeout, because it'll have only 5 seconds left to do its job.
- As calls are performed independently and  sequentially, `Pid_Remote_20` will also time-out.

And the question is what semantics should that timeout argument have:
1. `delegate:call/3` will be doing everything sequentially and will allow every gen_server to spend no more than timeout doing it's job. In that case we'll have only  some loose upper bound on amount of time `delegate:call/3` will take - `Timeout * ( MaxNumberGenServerPerRemoteNode + LocalGenServerCount)`. It's a completely insane API design.
2. `delegate:call/3` should return no later than after timeout. We are doing everything sequentially, and every next gen_server will have less and less time to handle the request. Reasonable for consumers of `delegate` API, but still very suboptimal.
3. `delegate:call/3` will be doing everything in parallel (going to nodes, spreading work to gen_servers on that nodes). And only in that case the timeout argument makes perfect sense: `delegate:call/3` should return after that amount of time has passed, every gen_server in the list will have the same amount of time to do its job. Making `delegate` more asynchronous will be just a side-effect of making `delegate:call` API sane.

**Comment by Ayanda-D** [comment]
I think the implications on queues when delegates do timeout are just too extreme. With the duration of a queue to complete processing being nondeterministic, I think its best leave out ANY queue related calls using `infinity`, despite quest to ensure cross-node timeouts "in as many places as possible". Also don't think there's any timeout policy we can define to entirely fulfil any requirement placed on a queue, even if it means re-implementing delegates, or improving them in any manner. Retaining `infinity` is safest, and we put emphasis of introducing timeouts only to cross-node calls to non-service affecting parts of the system.

**Comment by michaelklishin** [comment]
@Ayanda-D fair enough. Delegates are almost exclusively used to interact with queues, so I'm going to close this for now. Possibly a better solution will emerge in the future. Decisions are still progress.

@binarin please feel free to submit a PR with the delegate throughput improvements you have in mind. Keep in mind that we are not looking to introduce delegate timeouts any more, at least for now.
